File: project/src/nwr/gui/controls/TextBox.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package nwr.gui.controls;

import jzrlib.core.ExtList;
import jzrlib.utils.Logger;
import jzrlib.core.Rect;
import nwr.engine.BaseControl;
import nwr.engine.BaseScreen;
import nwr.engine.ControlStyles;
import nwr.engine.KeyEventArgs;
import nwr.engine.MouseEventArgs;
import nwr.engine.MouseWheelEventArgs;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public class TextBox extends TextControl
{
    private static final int LK_INDEP = 0;
    private static final int LK_FIRST = 1;
    private static final int LK_SECOND = 2;

    private static class HyperLink
    {
        public int symBeg;
        public int symEnd;
        public int Row;
        public int x1;
        public int x2;
        public int Kind;
    }

    private boolean fLinks;
    private final ExtList<HyperLink> fLinksList;
    private final ScrollBar fScrollBar;
    private int fTopIndex;
    
    public ILinkClickEvent OnLinkClick;

    public boolean LastVisible;
    public int LinkColor;

    public TextBox(BaseControl owner)
    {
        super(owner);
        this.fTopIndex = 0;

        this.fScrollBar = new ScrollBar(this);
        this.fScrollBar.setMin(0);
        this.fScrollBar.setMax(0);
        this.fScrollBar.setPos(0);
        this.fScrollBar.OnChange = this::OnScroll;
        this.fScrollBar.setLeft(super.getWidth() - this.fScrollBar.getWidth());
        this.fScrollBar.setTop(0);
        this.fScrollBar.setHeight(104);

        this.fLinks = false;
        this.fLinksList = new ExtList<>(true);

        super.setHeight(100);
        super.setWidth(200);
        this.LinkColor = BaseScreen.clSkyBlue;
        super.fLineHeight = 20;
    }

    @Override
    protected void dispose(boolean disposing)
    {
        if (disposing) {
            this.fLinksList.dispose();
        }
        super.dispose(disposing);
    }

    public final boolean getLinks()
    {
        return this.fLinks;
    }

    public final int getPageSize()
    {
        return this.fVisibleLines;
    }

    public final int getTopIndex()
    {
        return this.fTopIndex;
    }

    public final void setLinks(boolean Value)
    {
        this.fLinks = Value;
    }

    public final void setTopIndex(int Value)
    {
        if (this.fTopIndex != Value) {
            this.fTopIndex = Value;
            this.fScrollBar.setPos(this.fTopIndex);
        }
    }

    @Override
    protected void doKeyDownEvent(KeyEventArgs eventArgs)
    {
        switch (eventArgs.Key) {
            case GK_UP:
                if (this.fTopIndex > 0) {
                    this.setTopIndex(this.getTopIndex() - 1);
                }
                break;

            case GK_DOWN:
                if (this.fTopIndex < this.fLines.getCount() - 1 && this.fTopIndex + this.fVisibleLines < this.fLines.getCount()) {
                    this.setTopIndex(this.getTopIndex() + 1);
                }
                break;

            case GK_HOME:
                this.setTopIndex(0);
                break;

            case GK_END: {
                int idx = this.fLines.getCount() - this.fVisibleLines;
                if (idx >= 0) {
                    this.setTopIndex(idx);
                }
                break;
            }

            case GK_PRIOR:
                if (this.fTopIndex - this.fVisibleLines < 0) {
                    this.setTopIndex(0);
                } else {
                    this.setTopIndex(this.getTopIndex() - this.fVisibleLines);
                }
                break;

            case GK_NEXT:
                if (this.fTopIndex + this.fVisibleLines > this.fLines.getCount() - 1) {
                    int idx = this.fLines.getCount() - this.fVisibleLines;
                    if (idx >= 0) {
                        this.setTopIndex(idx);
                    }
                } else {
                    this.setTopIndex(this.getTopIndex() + this.fVisibleLines);
                }
                break;
        }

        super.doKeyDownEvent(eventArgs);
    }

    @Override
    protected void doMouseWheelEvent(MouseWheelEventArgs eventArgs)
    {
        super.doMouseWheelEvent(eventArgs);

        switch (eventArgs.WheelDelta) {
            case -1:
                if (this.fTopIndex > 0) {
                    this.setTopIndex(this.getTopIndex() - 1);
                }
                break;

            case +1:
                if (this.fTopIndex < this.fLines.getCount() - 1 && this.fTopIndex + this.fVisibleLines < this.fLines.getCount()) {
                    this.setTopIndex(this.getTopIndex() + 1);
                }
                break;
        }
    }

    private int findFirstLink(int row)
    {
        int num = this.fLinksList.getCount();
        for (int i = 0; i < num; i++) {
            HyperLink link = this.fLinksList.get(i);
            if (link.Row == row) {
                return i;
            }
        }
        return -1;
    }
    
    @Override
    protected void doPaintLine(BaseScreen screen, int index, Rect rect)
    {
        try {
            if (index < 0 || index >= super.fLines.getCount()) {
                return;
            }

            int x = rect.Left;
            int y = rect.Top;
            String line = super.fLines.get(index);
            screen.setTextColor(this.TextColor, true);

            if (!this.fLinks) {
                screen.drawText(x, y, line, 0);
            } else {
                int lnkIndex = this.findFirstLink(index);
                int offset = x;

                if (lnkIndex < 0) {
                    screen.drawText(x, y, line, 0);
                } else {
                    String chunk = "";
                    HyperLink curLink = null;

                    int i = 0;
                    while (i < line.length()) {
                        char chr = line.charAt(i);

                        if (chr != '@') {
                            chunk += chr;
                        } else {
                            if (curLink == null) {
                                screen.setTextColor(this.TextColor, true);
                                screen.drawText(offset, y, chunk, 0);

                                int sw = screen.getTextWidth(chunk);
                                offset += sw;

                                curLink = this.fLinksList.get(lnkIndex);
                                curLink.x1 = offset;
                            } else {
                                screen.setTextColor(this.LinkColor, true);
                                screen.drawText(offset, y, chunk, 0);

                                int sw = screen.getTextWidth(chunk);
                                offset += sw;

                                curLink.x2 = offset;
                                curLink = null;

                                lnkIndex++;
                            }

                            chunk = "";
                        }
                        
                        i++;
                    }

                    // end chunk
                    if (chunk.length() != 0) {
                            if (curLink == null) {
                                screen.setTextColor(this.TextColor, true);
                                screen.drawText(offset, y, chunk, 0);
                            } else {
                                screen.setTextColor(this.LinkColor, true);
                                screen.drawText(offset, y, chunk, 0);

                                int sw = screen.getTextWidth(chunk);
                                offset += sw;

                                curLink.x2 = offset;
                                curLink = null;
                            }
                    }
                }
            }
        } catch (Exception ex) {
            Logger.write("TextBox.DoPaintLine(): " + ex.getMessage());
        }
    }

    @Override
    protected void doPaintEvent(BaseScreen screen)
    {
        try {
            if (super.getFont() == null) {
                return;
            }

            Rect rt = super.getClientRect();
            if ((super.ControlStyle.contains(ControlStyles.csOpaque))) {
                screen.fillRect(rt, BaseScreen.clBlack);
            }
            CtlCommon.drawCtlBorder(screen, rt);

            screen.setTextColor(BaseScreen.clGold, true);

            int lb = this.fTopIndex;
            int hb = super.fLines.getCount() - 1;
            if (hb - lb >= this.fVisibleLines) {
                hb = lb + (this.fVisibleLines - 1);
            }

            rt = super.getIntRect();
            Rect r2 = new Rect();
            r2.Left = rt.Left + super.Margin;
            r2.Right = rt.Right - super.Margin;
            int top = rt.Top + super.Margin;

            for (int i = lb; i <= hb; i++) {
                r2.Top = top + (i - this.fTopIndex) * this.fLineHeight;
                r2.Bottom = r2.Top + this.fLineHeight - super.Margin;
                this.doPaintLine(screen, i, r2);
            }
        } catch (Exception ex) {
            Logger.write("TextBox.DoPaintTo(): " + ex.getMessage());
        }
    }

    @Override
    public Rect getIntRect()
    {
        Rect result = super.getIntRect();
        if (this.fScrollBar != null) {
            result.Right = result.Right - this.fScrollBar.getWidth() - (this.Margin << 1);
        }
        return result;
    }

    @Override
    protected void doMouseDownEvent(MouseEventArgs eventArgs)
    {
        super.doMouseDownEvent(eventArgs);

        Rect r = super.getIntRect();
        int eX = eventArgs.X - r.Left;
        int eY = eventArgs.Y - r.Top;

        int row = this.fTopIndex + eY / this.fLineHeight;
        if (row >= 0 && row < super.fLines.getCount()) {
            int num = this.fLinksList.getCount();
            for (int i = 0; i < num; i++) {
                HyperLink link = this.fLinksList.get(i);

                if (link.Row == row && (eX > link.x1 && eX < link.x2)) {
                    this.doLinkClickEvent(link, i);
                    return;
                }
            }
        }
    }

    private void doLinkClickEvent(HyperLink link, int idx)
    {
        if (this.OnLinkClick != null) {
            String linkValue = this.processLink(link, idx);
            this.OnLinkClick.invoke(this, linkValue);
        }
    }

    private String processLink(HyperLink link, int idx)
    {
        String id = "";

        String line;
        switch (link.Kind) {
            case LK_INDEP:
                line = this.fLines.get(link.Row);
                id = line.substring(link.symBeg + 1, link.symEnd);
                break;

            case LK_FIRST: {
                line = this.fLines.get(link.Row);
                id = line.substring(link.symBeg + 1, link.symEnd + 1);

                HyperLink nextLink = this.fLinksList.get(idx + 1);
                line = this.fLines.get(nextLink.Row);
                id = id + line.substring(nextLink.symBeg, nextLink.symEnd);
                break;
            }

            case LK_SECOND: {
                HyperLink prevLink = this.fLinksList.get(idx - 1);
                line = this.fLines.get(prevLink.Row);
                id = line.substring(prevLink.symBeg + 1, prevLink.symEnd + 1);

                line = this.fLines.get(link.Row);
                id = id + line.substring(link.symBeg, link.symEnd);
                break;
            }
        }
        
        return id;
    }

    @Override
    protected void prepareText()
    {
        if (this.fLinks) {
            this.fLinksList.clear();

            HyperLink curLink = null;
            for (int row = 0; row < this.fLines.getCount(); row++) {
                String text = this.fLines.get(row);
                if (text.length() < 1) {
                    continue;
                }

                int idx = 0;
                while (idx < text.length()) {
                    char c = text.charAt(idx);
                    if (c == '@') {
                        if (curLink == null) {
                            curLink = new HyperLink();
                            curLink.Row = row;
                            curLink.Kind = LK_INDEP;
                            curLink.symBeg = idx;
                            this.fLinksList.add(curLink);
                        } else {
                            curLink.symEnd = idx;
                            curLink = null;
                        }
                    }
                    idx++;
                }

                if (curLink != null) {
                    curLink.symEnd = text.length() - 1;
                    curLink.Kind = LK_FIRST;

                    curLink = new HyperLink();
                    curLink.Row = row + 1;
                    curLink.Kind = LK_SECOND;
                    curLink.symBeg = 0;
                    this.fLinksList.add(curLink);
                }
            }
        }
    }

    @Override
    protected void doResizeEvent()
    {
        super.doResizeEvent();
        if (this.fScrollBar != null) {
            this.fScrollBar.setLeft(super.getWidth() - this.fScrollBar.getWidth());
            this.fScrollBar.setHeight(super.getHeight());
        }
    }

    @Override
    protected void OnChange(Object Sender)
    {
        super.OnChange(Sender);

        if (this.fScrollBar != null) {
            if (this.fLines.getCount() == 0) {
                this.fScrollBar.setMax(0);
                this.fScrollBar.setPos(0);
            } else {
                this.fScrollBar.setMax(this.fLines.getCount() - this.fVisibleLines);
                this.fScrollBar.setPos(0);
            }
        }

        if (this.LastVisible) {
            if (super.fLines.getCount() - this.getPageSize() < 0) {
                this.setTopIndex(0);
            } else {
                this.setTopIndex(super.fLines.getCount() - this.getPageSize());
            }
        }
    }

    protected final void OnScroll(Object Sender)
    {
        this.fTopIndex = this.fScrollBar.getPos();
    }

    public boolean isValidLink(String linkValue)
    {
        return true;
    }
}

File: project/GUI/Controls/TextBox.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih.
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using BSLib;
using ZRLib.Core;
using ZRLib.Engine;

namespace NWR.GUI.Controls
{
    public delegate void ILinkClickEvent(object sender, string linkValue);

    public class TextBox : TextControl
    {
        private const int LK_INDEP = 0;
        private const int LK_FIRST = 1;
        private const int LK_SECOND = 2;

        private class HyperLink
        {
            public int SymBeg;
            public int SymEnd;
            public int Row;
            public int X1;
            public int X2;
            public int Kind;
        }

        private bool fLinks;
        private readonly ExtList<HyperLink> fLinksList;
        private readonly ScrollBar fScrollBar;
        private int fTopIndex;

        public ILinkClickEvent OnLinkClick;

        public bool LastVisible;
        public int LinkColor;

        public TextBox(BaseControl owner)
            : base(owner)
        {
            fTopIndex = 0;

            fScrollBar = new ScrollBar(this);
            fScrollBar.Min = 0;
            fScrollBar.Max = 0;
            fScrollBar.Pos = 0;
            fScrollBar.OnChange = OnScroll;
            fScrollBar.Left = Width - fScrollBar.Width;
            fScrollBar.Top = 0;
            fScrollBar.Height = 104;

            fLinks = false;
            fLinksList = new ExtList<HyperLink>(true);

            Height = 100;
            Width = 200;
            LinkColor = Colors.SkyBlue;
            fLineHeight = 20;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing) {
                fLinksList.Dispose();
            }
            base.Dispose(disposing);
        }

        public bool Links
        {
            get {
                return fLinks;
            }
            set {
                fLinks = value;
            }
        }

        public int PageSize
        {
            get {
                return fVisibleLines;
            }
        }

        public int TopIndex
        {
            get {
                return fTopIndex;
            }
            set {
                if (fTopIndex != value) {
                    fTopIndex = value;
                    fScrollBar.Pos = fTopIndex;
                }
            }
        }



        protected override void DoKeyDownEvent(KeyEventArgs eventArgs)
        {
            switch (eventArgs.Key) {
                case Keys.GK_UP:
                    if (fTopIndex > 0) {
                        TopIndex = TopIndex - 1;
                    }
                    break;

                case Keys.GK_DOWN:
                    if (fTopIndex < fLines.Count - 1 && fTopIndex + fVisibleLines < fLines.Count) {
                        TopIndex = TopIndex + 1;
                    }
                    break;

                case Keys.GK_HOME:
                    TopIndex = 0;
                    break;

                case Keys.GK_END:
                    {
                        int idx = fLines.Count - fVisibleLines;
                        if (idx >= 0) {
                            TopIndex = idx;
                        }
                        break;
                    }

                case Keys.GK_PRIOR:
                    if (fTopIndex - fVisibleLines < 0) {
                        TopIndex = 0;
                    } else {
                        TopIndex = TopIndex - fVisibleLines;
                    }
                    break;

                case Keys.GK_NEXT:
                    if (fTopIndex + fVisibleLines > fLines.Count - 1) {
                        int idx = fLines.Count - fVisibleLines;
                        if (idx >= 0) {
                            TopIndex = idx;
                        }
                    } else {
                        TopIndex = TopIndex + fVisibleLines;
                    }
                    break;
            }

            base.DoKeyDownEvent(eventArgs);
        }

        protected override void DoMouseWheelEvent(MouseWheelEventArgs eventArgs)
        {
            base.DoMouseWheelEvent(eventArgs);

            switch (eventArgs.WheelDelta) {
                case -1:
                    if (fTopIndex > 0) {
                        TopIndex = TopIndex - 1;
                    }
                    break;

                case +1:
                    if (fTopIndex < fLines.Count - 1 && fTopIndex + fVisibleLines < fLines.Count) {
                        TopIndex = TopIndex + 1;
                    }
                    break;
            }
        }

        private int FindFirstLink(int row)
        {
            int num = fLinksList.Count;
            for (int i = 0; i < num; i++) {
                HyperLink link = fLinksList[i];
                if (link.Row == row) {
                    return i;
                }
            }
            return -1;
        }

        protected override void DoPaintLine(BaseScreen screen, int index, ExtRect rect)
        {
            try {
                if (index < 0 || index >= fLines.Count) {
                    return;
                }

                int x = rect.Left;
                int y = rect.Top;
                string line = fLines[index];
                screen.SetTextColor(TextColor, true);

                if (!fLinks) {
                    screen.DrawText(x, y, line, 0);
                } else {
                    int lnkIndex = FindFirstLink(index);
                    int offset = x;

                    if (lnkIndex < 0) {
                        screen.DrawText(x, y, line, 0);
                    } else {
                        string chunk = "";
                        HyperLink curLink = null;

                        int i = 0;
                        while (i < line.Length) {
                            char chr = line[i];

                            if (chr != '@') {
                                chunk += chr;
                            } else {
                                if (curLink == null) {
                                    screen.SetTextColor(TextColor, true);
                                    screen.DrawText(offset, y, chunk, 0);

                                    int sw = screen.GetTextWidth(chunk);
                                    offset += sw;

                                    curLink = fLinksList[lnkIndex];
                                    curLink.X1 = offset;
                                } else {
                                    screen.SetTextColor(LinkColor, true);
                                    screen.DrawText(offset, y, chunk, 0);

                                    int sw = screen.GetTextWidth(chunk);
                                    offset += sw;

                                    curLink.X2 = offset;
                                    curLink = null;

                                    lnkIndex++;
                                }

                                chunk = "";
                            }

                            i++;
                        }

                        // end chunk
                        if (chunk.Length != 0) {
                            if (curLink == null) {
                                screen.SetTextColor(TextColor, true);
                                screen.DrawText(offset, y, chunk, 0);
                            } else {
                                screen.SetTextColor(LinkColor, true);
                                screen.DrawText(offset, y, chunk, 0);

                                int sw = screen.GetTextWidth(chunk);
                                offset += sw;

                                curLink.X2 = offset;
                                curLink = null;
                            }
                        }
                    }
                }
            } catch (Exception ex) {
                Logger.Write("TextBox.DoPaintLine(): " + ex.Message);
            }
        }

        protected override void DoPaintEvent(BaseScreen screen)
        {
            try {
                if (Font == null) {
                    return;
                }

                ExtRect rt = ClientRect;
                if ((ControlStyle.Contains(ControlStyles.сsOpaque))) {
                    screen.FillRect(rt, Colors.Black);
                }
                CtlCommon.DrawCtlBorder(screen, rt);

                screen.SetTextColor(Colors.Gold, true);

                int lb = fTopIndex;
                int hb = fLines.Count - 1;
                if (hb - lb >= fVisibleLines) {
                    hb = lb + (fVisibleLines - 1);
                }

                rt = base.IntRect;
                ExtRect r2 = new ExtRect();
                r2.Left = rt.Left + Margin;
                r2.Right = rt.Right - Margin;
                int top = rt.Top + Margin;

                for (int i = lb; i <= hb; i++) {
                    r2.Top = top + (i - fTopIndex) * fLineHeight;
                    r2.Bottom = r2.Top + fLineHeight - Margin;
                    DoPaintLine(screen, i, r2);
                }
            } catch (Exception ex) {
                Logger.Write("TextBox.DoPaintTo(): " + ex.Message);
            }
        }

        public override ExtRect IntRect
        {
            get {
                ExtRect result = base.IntRect;
                if (fScrollBar != null) {
                    result.Right = result.Right - fScrollBar.Width - (Margin << 1);
                }
                return result;
            }
        }

        protected override void DoMouseDownEvent(MouseEventArgs eventArgs)
        {
            base.DoMouseDownEvent(eventArgs);

            ExtRect r = base.IntRect;
            int eX = eventArgs.X - r.Left;
            int eY = eventArgs.Y - r.Top;

            int row = fTopIndex + eY / fLineHeight;
            if (row >= 0 && row < fLines.Count) {
                int num = fLinksList.Count;
                for (int i = 0; i < num; i++) {
                    HyperLink link = fLinksList[i];

                    if (link.Row == row && (eX > link.X1 && eX < link.X2)) {
                        DoLinkClickEvent(link, i);
                        return;
                    }
                }
            }
        }

        private void DoLinkClickEvent(HyperLink link, int idx)
        {
            if (OnLinkClick != null) {
                string linkValue = ProcessLink(link, idx);
                OnLinkClick.Invoke(this, linkValue);
            }
        }

        private string ProcessLink(HyperLink link, int idx)
        {
            string id = "";

            string line;
            switch (link.Kind) {
                case LK_INDEP:
                    line = fLines[link.Row];
                    id = line.Substring(link.SymBeg + 1, link.SymEnd - (link.SymBeg + 1));
                    break;

                case LK_FIRST:
                    {
                        line = fLines[link.Row];
                        id = line.Substring(link.SymBeg + 1, link.SymEnd + 1 - (link.SymBeg + 1));

                        HyperLink nextLink = fLinksList[idx + 1];
                        line = fLines[nextLink.Row];
                        id = id + line.Substring(nextLink.SymBeg, nextLink.SymEnd - nextLink.SymBeg);
                        break;
                    }

                case LK_SECOND:
                    {
                        HyperLink prevLink = fLinksList[idx - 1];
                        line = fLines[prevLink.Row];
                        id = line.Substring(prevLink.SymBeg + 1, prevLink.SymEnd + 1 - (prevLink.SymBeg + 1));

                        line = fLines[link.Row];
                        id = id + line.Substring(link.SymBeg, link.SymEnd - link.SymBeg);
                        break;
                    }
            }

            return id;
        }

        protected override void PrepareText()
        {
            if (fLinks) {
                fLinksList.Clear();

                HyperLink curLink = null;
                for (int row = 0; row < fLines.Count; row++) {
                    string text = fLines[row];
                    if (text.Length < 1) {
                        continue;
                    }

                    int idx = 0;
                    while (idx < text.Length) {
                        char c = text[idx];
                        if (c == '@') {
                            if (curLink == null) {
                                curLink = new HyperLink();
                                curLink.Row = row;
                                curLink.Kind = LK_INDEP;
                                curLink.SymBeg = idx;
                                fLinksList.Add(curLink);
                            } else {
                                curLink.SymEnd = idx;
                                curLink = null;
                            }
                        }
                        idx++;
                    }

                    if (curLink != null) {
                        curLink.SymEnd = text.Length - 1;
                        curLink.Kind = LK_FIRST;

                        curLink = new HyperLink();
                        curLink.Row = row + 1;
                        curLink.Kind = LK_SECOND;
                        curLink.SymBeg = 0;
                        fLinksList.Add(curLink);
                    }
                }
            }
        }

        protected override void DoResizeEvent()
        {
            base.DoResizeEvent();
            if (fScrollBar != null) {
                fScrollBar.Left = Width - fScrollBar.Width;
                fScrollBar.Height = Height;
            }
        }

        protected override void OnChange(object Sender)
        {
            base.OnChange(Sender);

            if (fScrollBar != null) {
                if (fLines.Count == 0) {
                    fScrollBar.Max = 0;
                    fScrollBar.Pos = 0;
                } else {
                    fScrollBar.Max = fLines.Count - fVisibleLines;
                    fScrollBar.Pos = 0;
                }
            }

            if (LastVisible) {
                if (fLines.Count - PageSize < 0) {
                    TopIndex = 0;
                } else {
                    TopIndex = fLines.Count - PageSize;
                }
            }
        }

        protected void OnScroll(object Sender)
        {
            fTopIndex = fScrollBar.Pos;
        }

        public virtual bool IsValidLink(string linkValue)
        {
            return true;
        }
    }
}
